crash:  invalid memory address or nil pointer dereference [1]

When hitting "https://127.0.0.1:8080/ws" the pool is erroring out . You might need to request "https://127.0.0.1:8080/ws" multiple times to get this.


```

2019-07-26 xx:xx:06.435 [TRC] POOL: Share PPLNS percentages are: (map[string]*big.Rat) (len=2) {
 (string) (len=64) "7aea0a63bb7099e256a5dd367ebdca0f8aaf4332745881d4efa7d339a94802c1": (*big.Rat)(0xc00080f1c0)(7/33),
 (string) (len=64) "193c4b8fd02aaed33ab9c5418ace9bec4047f61f923767bceb5a51c6e368bfa6": (*big.Rat)(0xc00080f0c0)(26/33)
}

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x808326]

goroutine 42 [running]:
github.com/gorilla/websocket.(*Conn).prepWrite(0x0, 0x1, 0xc0001cbd58, 0x40d4e9)
        /home/r00t/go/pkg/mod/github.com/gorilla/websocket@v1.2.0/conn.go:467 +0x26
github.com/gorilla/websocket.(*Conn).NextWriter(0x0, 0x1, 0x30, 0xc000866900, 0x0, 0x0)
        /home/r00t/go/pkg/mod/github.com/gorilla/websocket@v1.2.0/conn.go:488 +0x39
github.com/gorilla/websocket.(*Conn).WriteJSON(0x0, 0xb60fc0, 0xc000866900, 0xc000866900, 0x0)
        /home/r00t/go/pkg/mod/github.com/gorilla/websocket@v1.2.0/json.go:22 +0x49
github.com/decred/dcrpool/gui.(*GUI).SendUpdatedValues(0xc0000ce280, 0xc00032fad0, 0xc000866870, 0x2, 0x2)
        /home/r00t/workdir/dcrpool/gui/websocket.go:62 +0x31d
github.com/decred/dcrpool/gui.(*GUI).Run.func2(0xc0000ce280, 0xe111e0, 0xc0004a2680)
        /home/r00t/workdir/dcrpool/gui/gui.go:283 +0x9d
created by github.com/decred/dcrpool/gui.(*GUI).Run
        /home/r00t/workdir/dcrpool/gui/gui.go:264 +0x75
```

--- a/dcrpool.go
+++ b/dcrpool.go
@@ -14,7 +14,6 @@ import (
 	"os/signal"
 	"runtime"
 
-	bolt "github.com/coreos/bbolt"
 	"github.com/decred/dcrd/dcrutil"
 	"github.com/decred/dcrd/rpcclient"
 
@@ -25,7 +24,6 @@ import (
 // miningPool represents a decred Proof-of-Work mining pool.
 type miningPool struct {
 	cfg     *config
-	db      *bolt.DB
 	httpc   *http.Client
 	ctx     context.Context
 	cancel  context.CancelFunc
@@ -146,7 +144,7 @@ func newPool(cfg *config) (*miningPool, error) {
 		MinerPorts:    minerPorts,
 	}
 
-	p.gui, err = gui.NewGUI(gcfg, p.hub, p.db)
+	p.gui, err = gui.NewGUI(gcfg, p.hub)
 	if err != nil {
 		return nil, err
 	}

--- a/gui/admin.go
+++ b/gui/admin.go
@@ -7,10 +7,8 @@ package gui
 import (
 	"html/template"
 	"net/http"
-	"strconv"
 	"strings"
 
-	bolt "github.com/coreos/bbolt"
 	"github.com/gorilla/csrf"
 
 	"github.com/decred/dcrpool/pool"
@@ -134,13 +132,7 @@ func (ui *GUI) PostBackup(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	err = ui.db.View(func(tx *bolt.Tx) error {
-		w.Header().Set("Content-Type", "application/octet-stream")
-		w.Header().Set("Content-Disposition", `attachment; filename="backup.db"`)
-		w.Header().Set("Content-Length", strconv.Itoa(int(tx.Size())))
-		_, err := tx.WriteTo(w)
-		return err
-	})
+	err = ui.hub.BackupDB(w)
 
 	if err != nil {
 		log.Errorf("Error backing up database: %v", err)

--- a/gui/gui.go
+++ b/gui/gui.go
@@ -20,7 +20,6 @@ import (
 
 	"golang.org/x/crypto/acme/autocert"
 
-	bolt "github.com/coreos/bbolt"
 	"github.com/gorilla/csrf"
 	"github.com/gorilla/mux"
 	"github.com/gorilla/sessions"
@@ -53,7 +52,6 @@ type Config struct {
 type GUI struct {
 	cfg         *Config
 	hub         *pool.Hub
-	db          *bolt.DB
 	templates   *template.Template
 	cookieStore *sessions.CookieStore
 	router      *mux.Router
@@ -116,11 +114,10 @@ func (ui *GUI) renderTemplate(w http.ResponseWriter, r *http.Request, name strin
 }
 
 // NewGUI creates an instance of the user interface.
-func NewGUI(cfg *Config, hub *pool.Hub, db *bolt.DB) (*GUI, error) {
+func NewGUI(cfg *Config, hub *pool.Hub) (*GUI, error) {
 	ui := &GUI{
 		cfg: cfg,
 		hub: hub,
-		db:  db,
 	}
 
 	switch cfg.ActiveNet.Name {

--- a/gui/websocket.go
+++ b/gui/websocket.go
@@ -1,9 +1,10 @@
 package gui
 
 import (
-	"github.com/decred/dcrpool/pool"
 	"net/http"
 	"time"
+
+	"github.com/decred/dcrpool/pool"
 
 	"github.com/gorilla/websocket"
 )
@@ -37,6 +38,7 @@ func (ui *GUI) RegisterWebSocket(w http.ResponseWriter, r *http.Request) {
 	ws, err := upgrader.Upgrade(w, r, nil)
 	if err != nil {
 		log.Errorf("websocket error: %v", err)
+		return
 	}
 
 	// Register our new client

--- a/pool/hub.go
+++ b/pool/hub.go
@@ -11,6 +11,7 @@ import (
 	"fmt"
 	"math/big"
 	"net/http"
+	"strconv"
 	"sync"
 	"sync/atomic"
 	"time"
@@ -1177,4 +1178,17 @@ func (h *Hub) CSRFSecret(getCSRF func() ([]byte, error)) ([]byte, error) {
 	}
 
 	return secret, nil
+}
+
+// BackupDB streams a backup of the database over an http response.
+func (h *Hub) BackupDB(w http.ResponseWriter) error {
+	err := h.db.View(func(tx *bolt.Tx) error {
+		w.Header().Set("Content-Type", "application/octet-stream")
+		w.Header().Set("Content-Disposition", `attachment; filename="backup.db"`)
+		w.Header().Set("Content-Length", strconv.Itoa(int(tx.Size())))
+		_, err := tx.WriteTo(w)
+		return err
+	})
+
+	return err
 }

